compositor_pipeline/src/queue/audio_queue.rs

```diff
@@ -0,0 +1,101 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    time::{Duration, Instant},
+};
+
+use super::QueueError;
+use compositor_render::{AudioSamplesBatch, AudioSamplesSet, InputId};
+
+#[derive(Debug)]
+pub struct AudioQueue {
+    /// Enqueued sample batches per output
+    input_queues: HashMap<InputId, VecDeque<AudioSamplesBatch>>,
+    /// Stream added timestamp offset in relation to clock start
+    timestamp_offsets: HashMap<InputId, Duration>,
+    /// Stream starting pts offsets (starting pts doesn't have to be 0)
+    start_pts: HashMap<InputId, Duration>,
+}
+
+impl AudioQueue {
+    pub fn new() -> Self {
+        AudioQueue {
+            input_queues: HashMap::new(),
+            timestamp_offsets: HashMap::new(),
+            start_pts: HashMap::new(),
+        }
+    }
+
+    pub fn add_input(&mut self, input_id: InputId) {
+        self.input_queues.insert(input_id, VecDeque::new());
+    }
+
+    pub fn remove_input(&mut self, input_id: &InputId) {
+        self.input_queues.remove(input_id);
+    }
+
+    pub fn enqueue_samples(
+        &mut self,
+        input_id: InputId,
+        mut samples_batch: AudioSamplesBatch,
+        clock_start: Instant,
+    ) -> Result<(), QueueError> {
+        let Some(input_queue) = self.input_queues.get_mut(&input_id) else {
```

Comment:
You are syncing here based on the first audio sample, have you considered syncing both video and audio on the first received packet of either of them?

Reply:
I have considered it, but since the video decoder deals with timestamps differently than the audio decoder, this wouldn't work. It comes down to the same conversation as in the previous suggestion - how the PTS of queue input should look like (should PTS for all inputs start at 0 and increase monotonically).